include/mbedtls/asn1.h

```diff
@@ -316,6 +316,58 @@ int mbedtls_asn1_get_sequence_of( unsigned char **p,
                           mbedtls_asn1_sequence *cur,
                           int tag);
 
+/**
+ * \brief                Traverse an ASN.1 SEQUENCE container and
+ *                       call a callback for each entry.
+ *
+ * \warning              This function is still experimental and may change
+ *                       at any time.
+ *
+ * \param p              The address of the pointer to the beginning of
+ *                       the ASN.1 SEQUENCE header. This is updated to
+ *                       point to the end of the ASN.1 SEQUENCE container
+ *                       on a successful invocation.
+ * \param end            The end of the ASN.1 SEQUENCE container.
+ * \param tag_must_mask  A mask to be applied to the ASN.1 tags found within
+ *                       the SEQUENCE before comparing to \p tag_must_value.
+ * \param tag_must_val   The required value of each ASN.1 tag found in the
+ *                       SEQUENCE, after masking with \p tag_must_mask.
+ *                       Mismatching tags lead to an error.
+ *                       For example, a value of \c 0 for both \p tag_must_mask
+ *                       and \p tag_must_val means that every tag is allowed,
+ *                       while a value of \c 0xFF for \p tag_must_mask means
+ *                       that \p tag_must_val is the only allowed tag.
+ * \param tag_may_mask   A mask to be applied to the ASN.1 tags found within
+ *                       the SEQUENCE before comparing to \p tag_may_value.
+ * \param tag_may_val    The desired value of each ASN.1 tag found in the
+ *                       SEQUENCE, after masking with \p tag_may_mask.
+ *                       Mismatching tags will be silently ignored.
+ *                       For example, a value of \c 0 for \p tag_may_mask and
+ *                       \p tag_may_val means that any tag will be considered,
+ *                       while a value of \c 0xFF for \p tag_may_mask means
+ *                       that all tags with value different from \p tag_may_val
+ *                       will be ignored.
+ * \param cb             The callback to trigger for each component
+ *                       in the ASN.1 SEQUENCE. If the callback returns
+ *                       a non-zero value, the function stops immediately,
+ *                       forwarding the callback's return value.
+ * \param ctx            The context to be passed to the callback \p cb.
+ *
+ * \return               \c 0 if successful the entire ASN.1 SEQUENCE
+ *                       was traversed without parsing or callback errors.
+ * \return               A negative ASN.1 error code on a parsing failure.
+ * \return               A non-zero error code forwarded from the callback
+ *                       \p cb in case the latter returns a non-zero value.
+ */
+int mbedtls_asn1_traverse_sequence_of(
+    unsigned char **p,
```

Comment:
Can the `unsigned char` be `const unsigned char`? I don't think we'd be modifying `**p`, only `p` or `*p`.

Reply:
`const unsigned char **` and `unsigned char **` do not mix. If this function used a different type, it couldn't be used with the other asn1 parsing functions.

Reply:
I agree with @Patater in that in principle the type of `p` should be `const unsigned char **` here. However, the rest of the ASN.1 parsing library unfortunately also uses `unsigned char **`, which we can't change and which forces the use of `unsigned char **` here, too (not just for consistency, but simply because this function calls e.g. `mbedtls_asn1_get_tag()` which has a `unsigned char **` argument). An alternative would be to extend the documentation of all ASN.1 parsing functions to explicitly mention that, despite the missing `const` qualification, the `**p` is never written to, and that it is hence safe to use the functions in a `const unsigned char **` context. If we did that, we could start improving the API by using `const unsigned char **` at least for this new API function, and doing a const-cast when calling the other parsing functions.

Reply:
We can tweak the API in Mbed TLS 3. Until we get there, let's keep the API consistent.

Reply:
Ok, no change here, then.